### PR TITLE
Implements simpleGet, simpleOptions, simpleHead, fourOhFour, directoryLinks user stories and refactors

### PR DESCRIPTION
**Pull request by VictorElizalde**

### CODE REVIEW

File: app/src/main/java/java_server/Routes.java [hunk 1]
```diff
@@ -33,4 +33,9 @@ public boolean isAnExistingFileInDirectory(String[] fileList, Request request) {
 
         return false;
     }
+
+    public FileResponse getHandler(Request request) {
+        if (isURIValid(request) && isAnExistingFileInDirectory(getDirectoryFileNames(), request)) return new FileResponse(directory, request.getURI());
```
Comment (jassa) [inline app/src/main/java/java_server/Routes.java]: no crees que se entendería mejor en varias líneas?

File: app/src/test/java/java_server/ResponseTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        request = new Request();
+    }
+
+    @Test
+    public void returns200OkStatusWhenSuccessfulGetRequestIsMade() throws Exception {
+        request.setHTTPMethod("GET");
+        request.setURI("file1");
+        response.getMessageBody(request);
+
+        Assert.assertEquals("200 OK", response.getStatusMessage(request));
```
Comment (jassa) [inline app/src/test/java/java_server/ResponseTest.java:28]: `Request` es una clase que representa un HTTP request – cuantas razones de cambio tiene?
`Response` cuántas razones de cambio tiene? Qué tipo de cambios futuros la podrían romper?

File: app/src/test/java/java_server/RoutesTest.java [hunk 3]
```diff
@@ -83,4 +93,28 @@ public void returnsANullPointerExceptionStringIfURIDoesntExistInDirectory() thro
 
         Assert.assertEquals(false, routes.isAnExistingFileInDirectory(fileList, request));
     }
+
```
Comment (jassa) [inline app/src/test/java/java_server/RoutesTest.java:96]: No es de este commit pero en la linea 66

**throwsNullPointerException**_WhenDirectoryDoesNotExist <- Ojo, NullPointerEx nunca es claro su origen, evitala

> Don’t just return an exception
> When returning an exception you should also provide a context of the error, why it happened, where it happened, source, location, etc. Mention the operation that failed and the type of failure.

https://victorelizalde.medium.com/objects-and-data-structures-error-handling-and-boundaries-in-clean-code-313472a6bec0

File: app/src/main/java/java_server/Routes.java [hunk 4]
```diff
@@ -1,27 +1,84 @@
...
 
+    private HashMap<String, HashMap<String, Responder>> getRoutesMap(Request request) {
+        routesMap.put("/", getRootMap());
+        routesMap.put("file1", getFileRouteMap(request));
+        routesMap.put("file2", getFileRouteMap(request));
+        routesMap.put("image.jpeg", getFileRouteMap(request));
+        routesMap.put("image.png", getFileRouteMap(request));
+        routesMap.put("image.gif", getFileRouteMap(request));
+        routesMap.put("text-file.txt", getFileRouteMap(request));
+        routesMap.put("logs", getLogRouteMap(request));
```
Comment (jassa) [inline app/src/main/java/java_server/Routes.java:28]: de esta forma tendrás que cambiar este mapa por cada archivo que quieras servir, cómo pudieras hacerlo OCP?

File: app/src/main/java/java_server/Routes.java [hunk 5]
```diff
@@ -1,27 +1,84 @@
...
+
+        return rootMap;
+    }
+
+    private HashMap<String, Responder> getFileRouteMap(Request request) {
+        fileRouteMap.put("GET", new FileResponder(directory, request.getURI()));
+        fileRouteMap.put("HEAD", new FileResponder(directory, request.getURI()));
+        fileRouteMap.put("OPTIONS", new FileResponder(directory, request.getURI()));
+        fileRouteMap.put("PUT", new FileResponder(directory, request.getURI()));
+        fileRouteMap.put("DELETE", new FileResponder(directory, request.getURI()));
```
Comment (jassa) [inline app/src/main/java/java_server/Routes.java:44]: evita repetición

File: app/src/main/java/java_server/Routes.java [hunk 6]
```diff
@@ -1,27 +1,84 @@
...
+
+
     public Routes(String directory) {
         this.directory = directory;
     }
 
+    private HashMap<String, HashMap<String, Responder>> getRoutesMap(Request request) {
+        routesMap.put("/", getRootMap());
+        routesMap.put("file1", getFileRouteMap(request));
+        routesMap.put("file2", getFileRouteMap(request));
```
Comment (jassa) [inline app/src/main/java/java_server/Routes.java:23]: generas este mapa de mapas para todas las rutas por cada request?

File: app/src/test/java/java_server/MethodOptionsResponderTest.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+public class MethodOptionsResponderTest {
+    @Test
+    public void returnsOPTIONSForRequest() throws Exception {
+        statusCode statusCode = new statusCode();
+        MethodOptionsResponder methodOptionsResponder = new MethodOptionsResponder("GET,HEAD,OPTIONS,PUT,DELETE");
+        byte[] fileBytes = methodOptionsResponder.getMessageBody();
+
+        Assert.assertTrue(new String(fileBytes).contains("GET,HEAD,OPTIONS,PUT,DELETE"));
```
Comment (jassa) [inline app/src/test/java/java_server/MethodOptionsResponderTest.java:14]: inconveniente tener que convertir bytes a string en esta prueba de alto nivel

File: app/src/test/java/java_server/MethodOptionsResponderTest.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        Assert.assertEquals("200 OK", methodOptionsResponder.getStatusCode(statusCode));
+    }
```
Comment (jassa) [inline app/src/test/java/java_server/MethodOptionsResponderTest.java:16]: con solo una prueba tienes confianza que funciona? las pruebas unitarias deben explorar edge cases; ningun option, un option, varios options, etc. Parece diseño "test after" y no TDD

File: app/src/test/java/java_server/ResponseParserTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    public void presentsTheResponseInAnOutputStream() throws Exception {
+        String response = "HTTP/1.1 200 OK\r\n" +
+                "Location: http://localhost:5000/\r\n" +
+                "Content-Type: text/html\r\n" +
+                "Allow: HEAD,DELETE,GET,OPTIONS,PUT\r\n" +
+                "Content-Length: 14\r\n" +
+                "\r\n" +
+                "file1 contents";
+
+        Assert.assertEquals(response, new String(responseParser.present(request)));
```
Comment (jassa) [inline app/src/test/java/java_server/ResponseParserTest.java:64]: "present" no revela muy bien intención

File: app/src/test/java/java_server/ResponseTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        Assert.assertEquals("Content-Type: image/jpeg", response.getContentType(request));
+    }
+
+    @Test
+    public void returnsTheByteLengthForFile1() throws Exception {
+        request.setHTTPMethod("GET");
+        request.setURI("file1");
+
+        Assert.assertEquals("Content-Length: 14", response.getContentLength(request));
+    }
```
Comment (jassa) [inline app/src/test/java/java_server/ResponseTest.java:69]: estos chequeos de mime type dependiendo de si el file es texto, html, jpg, etc... pertenecen a Response?

File: app/src/test/java/java_server/RoutesTest.java [hunk 11]
```diff
@@ -83,4 +93,28 @@ public void returnsANullPointerExceptionStringIfURIDoesntExistInDirectory() thro
...
+        Assert.assertEquals("HEAD,DELETE,GET,OPTIONS,PUT", routes.getOptions(request));
+    }
+
+    @Test
+    public void returnsRootHandlerWhenRootRequestIsMade() throws Exception {
+        request.setHTTPMethod("GET");
+        request.setURI("/");
+
+        Assert.assertThat(routes.getHandler(request), instanceOf(RootResponder.class));
+    }
```
Comment (jassa) [inline app/src/test/java/java_server/RoutesTest.java:119]: tus pruebas se ven bien, pero de verdad pertenecen todas a RoutesTest según SOLID?

File: default-server-views/index.html [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+<ul>
+    <li> <a href='/file1'>file1</a> </li>
+    <li> <a href='/file2'>file2</a> </li>
+    <li> <a href='/image.jpeg'>image.jpeg</a></li>
+    <li> <a href='/image.png'>image.png</a></li>
+    <li> <a href='/image.gif'>image.gif</a></li>
+    <li> <a href='/text-file.txt'>text-file.txt</a></li>
+</ul>
+</body>
+</html>
```
Comment (jassa) [inline default-server-views/index.html:14]: este index.html no debe ser estático; debe ser dinámico en base a los contenidos de un directorio

**Review by jassa — CHANGES_REQUESTED** [review]

File: app/src/test/java/java_server/RoutesTest.java (hunk 11 above)
Comment (VictorElizalde) [reply app/src/test/java/java_server/RoutesTest.java:119]: Segun yo las que estan ahorita si, porque todas usan algun metodo de routes y comparan el resultado, cuales estan extra o que no van de acuerdo a SOLID?

File: app/src/test/java/java_server/RoutesTest.java (hunk 3 above)
Comment (VictorElizalde) [reply app/src/test/java/java_server/RoutesTest.java:96]: Listo

File: app/src/main/java/java_server/Routes.java (hunk 5 above)
Comment (VictorElizalde) [reply app/src/main/java/java_server/Routes.java:44]: Listo

File: app/src/test/java/java_server/MethodOptionsResponderTest.java (hunk 7 above)
Comment (VictorElizalde) [reply app/src/test/java/java_server/MethodOptionsResponderTest.java:14]: Listo

File: app/src/test/java/java_server/MethodOptionsResponderTest.java (hunk 8 above)
Comment (VictorElizalde) [reply app/src/test/java/java_server/MethodOptionsResponderTest.java:16]: Listo

File: app/src/test/java/java_server/ResponseParserTest.java (hunk 9 above)
Comment (VictorElizalde) [reply app/src/test/java/java_server/ResponseParserTest.java:64]: Listo

File: app/src/test/java/java_server/RoutesTest.java (hunk 11 above)
Comment (jassa) [reply app/src/test/java/java_server/RoutesTest.java:119]: cada vez que agregues una ruta nueva, tu clase Routes cambia y tu prueba también; esto quiere decir que no quedó Open-Closed. Hay que encontrar una forma de generalizar el comportamiento de Routes para "extenderlo" sin modificarlo

File: default-server-views/index.html (hunk 12 above)
Comment (VictorElizalde) [reply default-server-views/index.html:14]: Ya lo hice pero no me esta corriendo el javascript
